src/diagnostics/diagnostics.jl

```diff
@@ -0,0 +1,681 @@
+#   can convert between the two and check if the diagnostics are well-posed in terms of the
+#   relationship between the periods and the timesteps. In some sense, you can think of
+#   ScheduledDiagnosticIterations as an internal representation and ScheduledDiagnosticTime
+#   as the external interface.
+#
+# - A function to convert a list of ScheduledDiagnosticIterations into a list of
+#   AtmosCallbacks. This function takes three arguments: the list of diagnostics and two
+#   dictionaries that map each scheduled diagnostic to an area of memory where to save the
+#   result and where to keep track of how many times the function was called (so that we
+#   can compute stuff like averages).
+#
+# - This file also also include several other files, including (but not limited to):
+#   - core_diagnostics.jl
+#   - turbconv_diagnostics.jl
+#   - default_diagnostics.jl (which defines all the higher-level interfaces and defaults)
+#   - reduction_identities.jl
+#   - diagnostic_utils.jl
+
+
+"""
+    DiagnosticVariable
+
+
+A recipe to compute a diagnostic variable from the state, along with some useful metadata.
+
+The primary use for `DiagnosticVariable`s is to be embedded in a `ScheduledDiagnostic` to
+compute diagnostics while the simulation is running.
+
+The metadata is used exclusively by the `output_writer` in the `ScheduledDiagnostic`. It is
+responsibility of the `output_writer` to follow the conventions about the meaning of the
+metadata and their use.
+
+In `ClimaAtmos`, we roughly follow the naming conventions listed in this file:
+https://docs.google.com/spreadsheets/d/1qUauozwXkq7r1g-L4ALMIkCNINIhhCPx
+
+Keyword arguments
+=================
+
+- `short_name`: Name used to identify the variable in the output files and in the file
+                names. Short but descriptive. `ClimaAtmos` follows the CMIP conventions and
+                the diagnostics are identified by the short name.
+
+- `long_name`: Name used to identify the variable in the output files.
+
+- `units`: Physical units of the variable.
+
+- `comments`: More verbose explanation of what the variable is, or comments related to how
+              it is defined or computed.
+
+- `compute_from_integrator!`: Function that compute the diagnostic variable from the state.
+                             It has to take two arguments: the `integrator`, and a
+                             pre-allocated area of memory where to write the result of the
+                             computation. It the no pre-allocated area is available, a new
+                             one will be allocated. To avoid extra allocations, this
+                             function should perform the calculation in-place (i.e., using
+                             `.=`).
+"""
+Base.@kwdef struct DiagnosticVariable{T}
+    short_name::String
+    long_name::String
+    units::String
+    comments::String
+    compute_from_integrator!::T
```

Comment:
The entire integrator cache is stored in `integrator.p`, maybe we can remove a few layers of complexity if we start off assuming that we have something like `integrator.p.X` where `X` are the components defined in this PR? That way, we can dispatch on them, and avoid need / entanglement of the integrator.

Reply:
I don't think I understand what you are proposing. Can you share some pseudo code of what you have in mind?

I do agree that we can remove the reference to the integrator in the definition of the diagnostic variable and take in `state`, and `params` directly. And I like the idea because it feel more natural and motivated to compute something from a physical state as opposed to an integrator, which should probably be considered an implementation detail. 

So, `compute_from_integrator!(out, integrator)` might become `!compute(out, state, params)`. Then, we call this function with the correct arguments when we set up the callbacks.

Reply:
> I do agree that we can remove the reference to the integrator in the definition of the diagnostic variable and take in `state`, and `params` directly. And I like the idea because it feel more natural and motivated to compute something from a physical state as opposed to an integrator, which should probably be considered an implementation detail.

I'm proposing exactly this. Basically, rather than having functions in here that take the integrator (`foo(integrator)`), we instead pass in the parts (e.g., `foo(u, p, t)`)